src/IssueViz/Editor/QuickActions/IssueLocationActionsSource.cs

```diff
@@ -0,0 +1,123 @@
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Language.Intellisense;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Tagging;
+using SonarLint.VisualStudio.IssueVisualization.Editor.LocationTagging;
+using SonarLint.VisualStudio.IssueVisualization.Editor.SelectedIssueTagging;
+using SonarLint.VisualStudio.IssueVisualization.Models;
+using SonarLint.VisualStudio.IssueVisualization.Selection;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Editor.QuickActions
+{
+    internal sealed class IssueLocationActionsSource : ISuggestedActionsSource
+    {
+        private readonly IAnalysisIssueSelectionService selectionService;
+        private readonly ITagAggregator<IIssueLocationTag> issueLocationsTagAggregator;
+        private readonly ITagAggregator<ISelectedIssueLocationTag> selectedIssueLocationsTagAggregator;
+
+        public IssueLocationActionsSource(IBufferTagAggregatorFactoryService bufferTagAggregatorFactoryService, ITextBuffer textBuffer, IAnalysisIssueSelectionService selectionService)
+        {
+            this.selectionService = selectionService;
+
+            issueLocationsTagAggregator = bufferTagAggregatorFactoryService.CreateTagAggregator<IIssueLocationTag>(textBuffer);
+            issueLocationsTagAggregator.TagsChanged += TagAggregator_TagsChanged;
+
+            selectedIssueLocationsTagAggregator = bufferTagAggregatorFactoryService.CreateTagAggregator<ISelectedIssueLocationTag>(textBuffer);
+            selectedIssueLocationsTagAggregator.TagsChanged += TagAggregator_TagsChanged;
+        }
+
+        public event EventHandler<EventArgs> SuggestedActionsChanged;
+
+        public IEnumerable<SuggestedActionSet> GetSuggestedActions(ISuggestedActionCategorySet requestedActionCategories, SnapshotSpan range, CancellationToken cancellationToken)
+        {
+            var allActions = new List<ISuggestedAction>();
+
+            if (IsOnIssueVisualization(range, out var issueVisualizations))
+            {
+                var actions = issueVisualizations.Select(x => new SelectIssueVisualizationAction(selectionService, x));
+                allActions.AddRange(actions);
+            } 
+
+            if (IsOnSelectedVisualization(range))
+            {
+                allActions.Add(new DeselectIssueVisualizationAction(selectionService));
+            }
+
+            if (allActions.Any())
+            {
+                return new[] {new SuggestedActionSet(allActions)};
+            }
+
+            return Enumerable.Empty<SuggestedActionSet>();
+        }
+
+        public Task<bool> HasSuggestedActionsAsync(ISuggestedActionCategorySet requestedActionCategories, SnapshotSpan range, CancellationToken cancellationToken)
+        {
+            return Task.Factory.StartNew(() => IsOnIssueVisualization(range, out _) || IsOnSelectedVisualization(range),
```

Comment:
Is this the recommended way to start new tasks in the VS threading cookbook?

Reply:
It is how they recommend it in the walkthrough
https://docs.microsoft.com/en-us/visualstudio/extensibility/walkthrough-displaying-light-bulb-suggestions?view=vs-2019#implement-the-isuggestedactionsource

Reply:
It's always work checking the VS threading cookbook, which was written by the platform devs and contains recommended best practices. It's less clear whether examples in docs follow best practices or not.
Unfortunately, the cookbook doesn't have a recommendation for calling synchronous code when implementing an async method.
Fingers crossed this is something that is checked by the VS async threading analyzers.